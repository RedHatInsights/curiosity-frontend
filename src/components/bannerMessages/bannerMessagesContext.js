import { useShallowCompareEffect } from 'react-use';
import { reduxActions, storeHooks } from '../../redux';
import { useProduct, useProductQuery } from '../productView/productViewContext';
import { dateHelpers } from '../../common';
import {
  rhsmConstants,
  RHSM_API_QUERY_GRANULARITY_TYPES as GRANULARITY_TYPES,
  RHSM_API_QUERY_SET_TYPES
} from '../../services/rhsm/rhsmConstants';

/**
 * Get app messages.
 *
 * @param {object} options
 * @param {Function} options.getMessageReports
 * @param {Function} options.useDispatch
 * @param {Function} options.useProduct
 * @param {Function} options.useProductQuery
 * @param {Function} options.useSelectorsResponse
 * @returns {object}
 */
const useGetAppMessages = ({
  getMessageReports = reduxActions.rhsm.getMessageReports,
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useProduct: useAliasProduct = useProduct,
  useProductQuery: useAliasProductQuery = useProductQuery,
  useSelectorsResponse: useAliasSelectorsResponse = storeHooks.reactRedux.useSelectorsResponse
} = {}) => {
  const { productId } = useAliasProduct();
  const query = useAliasProductQuery();
  const dispatch = useAliasDispatch();
  const { error, fulfilled, pending, data } = useAliasSelectorsResponse({
    id: 'messages',
    selector: ({ messages }) => messages?.report?.[productId]
  });

  useShallowCompareEffect(() => {
    if (productId) {
      const { startDate, endDate } = dateHelpers.getRangedDateTime('CURRENT');
      const updatedQuery = {
        ...query,
        [RHSM_API_QUERY_SET_TYPES.GRANULARITY]: GRANULARITY_TYPES.DAILY,
        [RHSM_API_QUERY_SET_TYPES.START_DATE]: startDate.toISOString(),
        [RHSM_API_QUERY_SET_TYPES.END_DATE]: endDate.toISOString()
      };

      getMessageReports(productId, updatedQuery)(dispatch);
    }
  }, [productId, query]);

  const updatedData = {
    cloudigradeMismatch: false
  };

  if (fulfilled) {
    const { messages = {} } = data || {};

    updatedData.cloudigradeMismatch =
      messages?.data
        ?.reverse()
        ?.find(
          ({ [rhsmConstants.RHSM_API_RESPONSE_TALLY_META_TYPES.HAS_CLOUDIGRADE_MISMATCH]: mismatch }) =>
            mismatch === true
        ) !== undefined;
  }

  return {
    error,
    fulfilled,
    pending,
    data: {
      ...updatedData
    }
  };
};

const context = {
  useGetAppMessages
};

export { context as default, context, useGetAppMessages };
