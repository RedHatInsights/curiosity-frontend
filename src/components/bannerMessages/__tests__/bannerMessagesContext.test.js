import { context, useBannerMessages, useRemoveBannerMessages, useSetBannerMessages } from '../bannerMessagesContext';

describe('BannerMessagesContext', () => {
  it('should return specific properties', () => {
    expect(context).toMatchSnapshot('specific properties');
  });

  it('should apply a hook for retrieving messages data from a selector', async () => {
    const { result } = await renderHook(() =>
      useBannerMessages({
        useSelector: () => [
          {
            id: 'lorem',
            title: 'ipsum'
          }
        ]
      })
    );

    expect(result).toMatchSnapshot('banner messages');
  });

  it('should apply a hook for retrieving messages data from a selector and apply new messages', async () => {
    const mockDispatch = jest.fn();
    const { result } = await renderHook(() =>
      useSetBannerMessages({
        useDispatch: () => mockDispatch,
        useProduct: () => ({ productId: 'dolorSit' }),
        useBannerMessages: () => [
          {
            id: 'lorem',
            title: 'ipsum'
          }
        ]
      })
    );

    result('new message');
    expect(mockDispatch.mock.calls).toMatchSnapshot('dispatch');
  });

  it('should apply a hook for retrieving messages data from a selector and remove messages', async () => {
    const mockDispatch = jest.fn();
    const { result } = await renderHook(() =>
      useRemoveBannerMessages({
        useDispatch: () => mockDispatch,
        useProduct: () => ({ productId: 'dolorSit' }),
        useBannerMessages: () => [
          {
            id: 'lorem',
            title: 'ipsum'
          }
        ]
      })
    );

    result('ipsum');
    expect(mockDispatch.mock.calls).toMatchSnapshot('dispatch');
  });
});
