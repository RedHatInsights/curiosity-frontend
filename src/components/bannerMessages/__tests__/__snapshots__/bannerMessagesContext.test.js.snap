// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BannerMessagesContext should apply a hook for retrieving messages data from a selectors: error response 1`] = `
Object {
  "data": Object {
    "cloudigradeMismatch": false,
  },
  "error": true,
  "fulfilled": undefined,
  "pending": undefined,
}
`;

exports[`BannerMessagesContext should apply a hook for retrieving messages data from a selectors: mock store error response 1`] = `
Object {
  "data": Object {
    "cloudigradeMismatch": false,
  },
  "error": true,
  "fulfilled": false,
  "pending": false,
}
`;

exports[`BannerMessagesContext should apply a hook for retrieving messages data from a selectors: mock store success response 1`] = `
Object {
  "data": Object {
    "cloudigradeMismatch": true,
  },
  "error": false,
  "fulfilled": true,
  "pending": false,
}
`;

exports[`BannerMessagesContext should apply a hook for retrieving messages data from a selectors: success response 1`] = `
Object {
  "data": Object {
    "cloudigradeMismatch": false,
  },
  "error": undefined,
  "fulfilled": true,
  "pending": undefined,
}
`;

exports[`BannerMessagesContext should return specific properties: specific properties 1`] = `
Object {
  "useGetAppMessages": [Function],
}
`;
